Fix jagged composite impl of flatten()

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115842
* #116111
* __->__ #115192

Need to handle this in `NestedTensor.__torch_function__()` since it's CompositeImplicit

jbschlosser committed:
Fix jagged composite impl of flatten()

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

cpuhrsch reviewed (approved):
Same comment about test, but otherwise stamped.

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -300,14 +301,24 @@ def _flatten_sig(input, start_dim=0, end_dim=-1):
+        product = functools.reduce(operator.mul, inp.shape[start_dim : end_dim + 1])
+        new_shape = (
+            list(inp.shape[:start_dim]) + [product] + list(inp.shape[end_dim + 1 :])
+        )
```
nit: `new_shape(*inp.shape[:start_dim], product, *inp.shape[end_dim + 1:])`

jbschlosser replied:
much nicer :0

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


TODO: tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


Need to handle this in `NestedTensor.__torch_function__()` since it's CompositeImplicit

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


Need to handle this in `NestedTensor.__torch_function__()` since it's CompositeImplicit

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Fix jagged composite impl of flatten()"


Need to handle this in `NestedTensor.__torch_function__()` since it's CompositeImplicit

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
